R/language_level_df.R

```diff
@@ -1,134 +1,101 @@
 #' @param LanguageTable data-frame of a cldf LanguageTable. Needs to have columns Glottocode and Language_ID or Language_level_ID.
-#' #' @param drop_question_marks logical vector indicating wether to turn ? into missing values (NA) or not.
+#' @param method character vector, choice between "singular_least_missing_data", "combine_random", "singular_random". combine_random = combine all datapoints for all the dialects/duplicates and if there is more than one datapoint for a given feature/word/variable choose randomly between them, singular_random = choose randomly between the dialects/duplicates, singular_least_missing_data = choose the dialect/duplicate which has the most datapoints.
 #' @return data-frame of ValueTable without duplicates
 #' @export
 
-# TODO SJG: can we fix the indentation here?
-# Hedvig response: Sure, to me that's a aesthtics thing, we can do it however you'd like.
+language_level_df <- function(ValueTable, LanguageTable,
+                              method = c("singular_least_missing_data", "combine_random", "singular_random"),
+                              drop_question_marks = TRUE) {
 
-
-language_level_df <- function(ValueTable = NULL,
-                              method = c( "singular_least_missing_data", "combine_random", "singular_random"),
-                              drop_question_marks = TRUE,
-                              LanguageTable = NULL){
-
-    if(length(method) != 1 & !(method %in% c( "singular_least_missing_data",
-                                              "combine_random", "singular_random"))){
+    if (length(method) != 1 & !(method %in% c("singular_least_missing_data", "combine_random", "singular_random"))) {
         stop("Method of merging is not defined.")
     }
 
     if (!all(c('Language_ID', 'Parameter_ID', 'Value') %in% colnames(ValueTable))) {
         stop("Invalid table format - ValueTable needs to have columns Language_ID/Parameter_ID/Value")
     }
 
-# The language level ID column is named different in different, CLDF datasets.
-# setting them to the same
-if(!("Language_level_ID" %in% colnames(LanguageTable))){
-    LanguageTable   <- LanguageTable %>%
-        dplyr::select(Language_ID = Glottocode, Language_level_ID = Language_ID)
-}
-
-if("Language_level_ID" %in% colnames(LanguageTable) & "ID" %in% colnames(LanguageTable)){
-    LanguageTable   <- LanguageTable %>%
-        dplyr::select(Language_ID = Glottocode, Language_level_ID)
-}
-
-# if there is a missing language level ID, which it can be in some datasets where only dialects get language level IDs and languages and families don't, then replace those with the content in the Language_ID column.
-LanguageTable   <- LanguageTable %>%
-  dplyr::mutate(Language_level_ID = ifelse(is.na(Language_level_ID)|
-                                               Language_level_ID == "", Language_ID, Language_level_ID))
-
-## QUESTION MARK ACTION
-
-#turn question marks into missing values if set to TRUE
```

Comment:
I think we should keep the part of this function that lets the user turn ? into NAs or not.

Reply:
Refactor means "tidy up and improve".

I removed the question mark thing as it is out of scope here and feels wrong for a few reasons: 
1. This is a method to make a language level data frame. Magically transforming data is  something else. 
2. It's better for people to operate on the original data and decide what to do with "?" explicitly. 
3. It's badly named (what's a question mark? "drop" means remove, but this recodes).
4. ideally rcldf should handle this at the CLDF level (https://github.com/SimonGreenhill/rcldf/issues/14)
5. It's excessive. If we add these sorts of features to the R package it's more code we need to maintain and we need to test. Users can do this easily using tools already available to them, so why do we need to put effort into implementation and maintenance:

```{r}
values <- gb$tables$ValueTable %>% mutate(Value = na_if(Value, "?"))

# or even:
lld <- gb$tables$ValueTable %>%
    mutate(Value = na_if(Value, "?")) %>%
    language_level_df(., gb$tables$LanguageTable, 'singular_least_missing_data')
```